tee-worker/omni-executor/rpc-server/src/methods/omni/get_approve_builder_fee_data.rs

```diff
@@ -0,0 +1,286 @@
+}
+
+#[derive(Serialize, Clone)]
+pub struct ApproveBuilderFeeData {
+	pub action: ApproveBuilderFeeAction,
+	pub nonce: u64,
+	pub signature: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct ApproveBuilderFeeAction {
+	#[serde(serialize_with = "serialize_hex")]
+	pub signature_chain_id: u64,
+	pub hyperliquid_chain: String,
+	pub max_fee_rate: String,
+	pub builder: Address,
+	pub nonce: u64,
+}
+
+fn is_testnet_chain(chain_id: ChainId) -> bool {
+	match chain_id {
+		// Testnet chain IDs
+		11155111 => true, // Ethereum Sepolia
+		97 => true,       // BSC Testnet
+		80001 => true,    // Polygon Mumbai
+		421614 => true,   // Arbitrum Sepolia
+		998 => true,      // HyperEVM Testnet
+		1337 => true,     // Local Anvil
+		31337 => true,    // Local Anvil
+		// Mainnet chain IDs
+		1 => false,     // Ethereum Mainnet
+		56 => false,    // BSC Mainnet
+		137 => false,   // Polygon Mainnet
+		42161 => false, // Arbitrum Mainnet
+		999 => false,   // HyperEVM Mainnet
+		_ => {
+			// Default to mainnet for unknown chain IDs
+			false
+		},
+	}
+}
+
+impl ApproveBuilderFeeAction {
+	fn domain(&self) -> Eip712Domain {
+		eip712_domain! {
+			name: "HyperliquidSignTransaction",
+			version: "1",
+			chain_id: self.signature_chain_id,
+			verifying_contract: Address::ZERO,
+		}
+	}
+
+	fn struct_hash(&self) -> B256 {
+		let items = (
+            keccak256("HyperliquidTransaction:ApproveBuilderFee(string hyperliquidChain,string maxFeeRate,address builder,uint64 nonce)"),
+            keccak256(&self.hyperliquid_chain),
+            keccak256(&self.max_fee_rate),
+            &self.builder,
+            &self.nonce
+        );
+		keccak256(items.abi_encode())
+	}
+
+	fn eip712_signing_hash(&self) -> B256 {
+		let mut digest_input = [0u8; 2 + 32 + 32];
+		digest_input[0] = 0x19;
+		digest_input[1] = 0x01;
+		digest_input[2..34].copy_from_slice(&self.domain().hash_struct()[..]);
+		digest_input[34..66].copy_from_slice(&self.struct_hash()[..]);
+		keccak256(digest_input)
+	}
+}
+
+pub fn register_get_approve_builder_fee_data(module: &mut RpcModule<RpcContext>) {
+	module
+		.register_async_method("omni_getApproveBuilderFeeData", |params, ctx, _| async move {
+			let params = params.parse::<GetApproveBuilderFeeDataParams>().map_err(|e| {
+				error!("Failed to parse params: {:?}", e);
+				ErrorCode::ParseError
+			})?;
+
+			let verification_code = match &params.user_auth {
+				UserAuth::Email(code) => code.clone(),
+				_ => {
+					error!("Only email authentication is supported for this method");
+					return Err(ErrorObject::from(ErrorCode::InvalidParams));
+				},
+			};
+
+			let email = match &params.user_id {
+				UserId::Email(email) => email.clone(),
+				_ => {
+					error!("User ID must be an email for this method");
+					return Err(ErrorObject::from(ErrorCode::ParseError));
+				},
+			};
+
+			verify_email_authentication(ctx.clone(), &params.client_id, &email, &verification_code)
```

Comment:
A dumb Q: do all these methods need email auth (code)? Or some can be done via agent wallet sig?

Either way we should be able to extract the auth verification process into a common function (maybe there's one already)